### PR TITLE
Implementation for #913 Paket.pack: add support for nuget dependencies conditional on target framework

### PR DESCRIPTION
**Pull request by bhugot**
Hi, this is a proposal for #913 

it's going to fix a bit of #2394 too

the template should be 'type  file'

and the syntax is the following 
````
type file
id My.Thing
authors Bob McBob
description
    A longer description
    on two lines.
version
    1.0
dependencies 
    framework: net45
        FSharp.Core 4.3.1
        My.OtherThing
    framework: netstandard11
        FSharp.Core 4.3.1  
````

### CODE REVIEW

**Comment by bhugot** [comment]
@forki Hi is it possible to review this one it's really a needed feature

File: src/Paket.Core/Packaging/NupkgWriter.fs [hunk 1]
```diff
@@ -83,23 +84,45 @@ module internal NupkgWriter =
             frameworkAssembliesList |> List.iter (buildFrameworkReferencesNode >> d.Add)
             metadataNode.Add d
 
-        let buildDependencyNode (Id, requirement:VersionRequirement) =
+        let buildDependencyNode (id, requirement:VersionRequirement) =
```
Comment (forki) [inline src/Paket.Core/Packaging/NupkgWriter.fs]: Can you please rename this thing. id is a function from the FSharp.Core and I think we should avoid confusion. Maybe name or idProperty or something like that

File: src/Paket.Core/Packaging/NupkgWriter.fs [hunk 2]
```diff
@@ -83,23 +84,45 @@ module internal NupkgWriter =
...
 
+        let buildGroupNode (framework:FrameworkIdentifier, add) = 
+            let g = XElement(ns + "group")
+            g.SetAttributeValue(XName.Get "targetFramework", framework.ToString())
+            add g
+            g
+
+
+        let buildDependencyNodes (excludedDependencies, add, dependencyList)  =
+            let takeFirstOfThree (a,_,_) = a
```
Comment (forki) [inline src/Paket.Core/Packaging/NupkgWriter.fs]: That's a nice trick, but could you please manually inline this? I don't think we really need these functions here. Thx

File: src/Paket.Core/Packaging/NupkgWriter.fs [hunk 3]
```diff
@@ -83,23 +84,45 @@ module internal NupkgWriter =
...
+
         let buildDependenciesNode excludedDependencies dependencyList =
             if List.isEmpty dependencyList then () else
             let d = XElement(ns + "dependencies")
-            dependencyList
-            |> List.filter (fun d -> Set.contains (fst d) excludedDependencies |> not)
-            |> List.iter (buildDependencyNode >> d.Add)
-            metadataNode.Add d
+            let groups = List.groupBy thirdOf3 dependencyList
+            if groups.Length = 1 &&  fst groups.Head = None then
```
Comment (forki) [inline src/Paket.Core/Packaging/NupkgWriter.fs]: Could you please use pattern matching here?

File: src/Paket.Core/Packaging/PackageMetaData.fs [hunk 4]
```diff
@@ -107,11 +107,14 @@ let (|Valid|Invalid|) md =
                 Symbols = s }
     | _ -> Invalid
 
-let addDependency (templateFile : TemplateFile) (dependency : PackageName * VersionRequirement) = 
+
+
+let addDependency (templateFile : TemplateFile) (dependency : PackageName * VersionRequirement * FrameworkIdentifier option) = 
     match templateFile with
     | CompleteTemplate(core, opt) ->
```
Comment (forki) [inline src/Paket.Core/Packaging/PackageMetaData.fs]: Same here. Please get rid of firstOf3

File: src/Paket.Core/PaketConfigFiles/TemplateFile.fs [hunk 5]
```diff
@@ -261,45 +262,80 @@ module internal TemplateFile =
...
+        | String.RemovePrefix "#" _ -> Some (Empty line)
+        | _ -> None    
+    type TargetFrameworkGroup = 
+        {
+            Framework : FrameworkIdentifier option
+            Dependencies : (PackageName * VersionRequirement * FrameworkIdentifier option) List}
+        static member ForNone = { Framework = None ; Dependencies = [] }
+        static member ForFramework framework = { Framework = framework ; Dependencies = [] }
+   
+    let private getDependencieByLine (fileName, lockFile:LockFile,currentVersion:SemVerInfo option, specificVersions:Map<string, SemVerInfo>, line:string, framework:FrameworkIdentifier option)=
```
Comment (forki) [inline src/Paket.Core/PaketConfigFiles/TemplateFile.fs]: Small typo in function name

File: src/Paket.Core/Versioning/FrameworkHandling.fs [hunk 6]
```diff
@@ -439,7 +439,6 @@ module FrameworkDetection =
 
     open Logging
     /// parse a string to construct a Netframework, NetCore, NetStandard, or other dotnet identifier
-    [<Obsolete "Use PlatformMatching.extractPlatforms instead">]
```
Comment (forki) [inline src/Paket.Core/Versioning/FrameworkHandling.fs]: ?

**Comment by bhugot** [comment]
Hope the change are good for you, my first time with F# so I don't know all the good way to do.

**Comment by forki** [comment]
It's really great contribution. Thank you. Only thing left is a bit of
documentation. You can find md files in /docs (or so). Sorry I am on
mobile. If that's done we can release as 5.1 (cc @matthid)

Am 18.06.2017 2:50 nachm. schrieb "Benjamin Hugot" <notifications@github.com

**Comment by bhugot** [comment]
Added documentation.

**Comment by forki** [comment]
thanks a lot! released in 5.1

**Comment by bhugot** [comment]
No problems it s helping my team too

**Comment by matthid** [comment]
Obviously I'm a bit late but shouldn't we be able to calculate the correct groups from the framework restrictions ourself? All we would need is the list of frameworks the package contains binaries for. Rest could be figured out by paket itself.

**Pull request merged**